### PR TITLE
[YQ-1997] OR REPLACE grammar support for EXTERNAL DATA SOURCE and EXTERNAL TABLE

### PR DESCRIPTION
**Pull request by Alnen**
### Changelog entry 

...

### Changelog category 

* New feature

### Additional information

...

### CODE REVIEW

File: ydb/core/kqp/gateway/utils/scheme_helpers.cpp [hunk 1]
```diff
@@ -61,12 +61,14 @@ bool SetDatabaseForLoginOperation(TString& result, bool getDomainLoginOnly, TMay
 
 void FillCreateExternalTableColumnDesc(NKikimrSchemeOp::TExternalTableDescription& externalTableDesc,
                                        const TString& name,
+                                       bool isReplace,
```
Comment (UgnineSirdis) [inline ydb/core/kqp/gateway/utils/scheme_helpers.cpp:64]: `replaceIfExists`

**Review by UgnineSirdis — CHANGES_REQUESTED** [review]

File: ydb/core/protos/flat_scheme_op.proto [hunk 2]
```diff
@@ -1806,6 +1806,7 @@ message TExternalTableDescription {
     optional string Location = 6;
     repeated TColumnDescription Columns = 7;
     optional bytes Content = 8;
+    optional bool ReplaceIfExists = 9;
```
Comment (UgnineSirdis) [inline ydb/core/protos/flat_scheme_op.proto:1809]: This field has no sense for describe operation. It's worth to write comment about it.

File: ydb/library/yql/sql/v1/format/sql_format.cpp [hunk 3]
```diff
@@ -891,27 +891,24 @@ friend struct TStaticData;
         Visit(msg.GetToken1());
         Visit(msg.GetBlock2());
         Visit(msg.GetBlock3());
-        Visit(msg.GetRule_simple_table_ref4());
-        Visit(msg.GetToken5());
+        Visit(msg.GetBlock4());
```
Comment (UgnineSirdis) [inline ydb/library/yql/sql/v1/format/sql_format.cpp:894]: We need to add tests for this new behaviour to `ydb/library/yql/sql/v1/format/sql_format_ut.cpp`

**Review by UgnineSirdis — CHANGES_REQUESTED** [review]

File: ydb/services/metadata/abstract/parsing.h [hunk 4]
```diff
@@ -13,6 +13,7 @@ namespace NYql {
 namespace NObjectOptionsParsing {
 Y_HAS_MEMBER(ExistingOk); // for create
 Y_HAS_MEMBER(MissingOk); // for drop
+Y_HAS_MEMBER(IsReplace); // for create
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/services/metadata/abstract/parsing.h]: Fixed

File: ydb/library/yql/sql/v1/node.h [hunk 5]
```diff
@@ -1231,7 +1231,7 @@ namespace NSQLTranslationV1 {
     TNodePtr BuildUpsertObjectOperation(TPosition pos, const TString& objectId, const TString& typeId,
         std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context);
     TNodePtr BuildCreateObjectOperation(TPosition pos, const TString& objectId, const TString& typeId,
-        bool existingOk, std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context);
+        bool existingOk, bool isReplace, std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context);
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/node.h]: Fixed

File: ydb/library/yql/sql/v1/object_processing.h [hunk 6]
```diff
@@ -40,19 +40,30 @@ class TCreateObject: public TObjectProcessorImpl {
     std::set<TString> FeaturesToReset;
 protected:
     bool ExistingOk = false;
+    bool IsReplace = false;
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/object_processing.h]: Fixed

File: ydb/library/yql/sql/v1/query.cpp [hunk 7]
```diff
@@ -1149,12 +1156,13 @@ class TCreateTableNode final: public TAstListNode {
     const TTableRef Table;
     const TCreateTableParameters Params;
     const bool ExistingOk;
+    const bool IsReplace;
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/query.cpp]: Fix

File: ydb/library/yql/sql/v1/query.cpp [hunk 8]
```diff
@@ -1149,12 +1156,13 @@ class TCreateTableNode final: public TAstListNode {
...
     TScopedStatePtr Scoped;
 };
 
-TNodePtr BuildCreateTable(TPosition pos, const TTableRef& tr, bool existingOk, const TCreateTableParameters& params, TScopedStatePtr scoped)
+TNodePtr BuildCreateTable(TPosition pos, const TTableRef& tr, bool existingOk, bool isReplace, const TCreateTableParameters& params, TScopedStatePtr scoped)
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/query.cpp]: Fix

File: ydb/library/yql/sql/v1/query.cpp [hunk 9]
```diff
@@ -2109,21 +2117,21 @@ class TDropRoles final: public TAstListNode {
 
 TNodePtr BuildUpsertObjectOperation(TPosition pos, const TString& objectId, const TString& typeId,
     std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context) {
-    return new TUpsertObject(pos, objectId, typeId, false, std::move(features), std::set<TString>(), context);
+    return new TUpsertObject(pos, objectId, typeId, false, false, std::move(features), std::set<TString>(), context);
 }
 TNodePtr BuildCreateObjectOperation(TPosition pos, const TString& objectId, const TString& typeId,
-    bool existingOk, std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context) {
-    return new TCreateObject(pos, objectId, typeId, existingOk, std::move(features), std::set<TString>(), context);
+    bool existingOk, bool isReplace, std::map<TString, TDeferredAtom>&& features, const TObjectOperatorContext& context) {
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/query.cpp]: Fixed

File: ydb/library/yql/sql/v1/source.h [hunk 10]
```diff
@@ -295,7 +295,7 @@ namespace NSQLTranslationV1 {
     TNodePtr BuildTopicKey(TPosition pos, const TDeferredAtom& cluster, const TDeferredAtom& name);
     TNodePtr BuildInputOptions(TPosition pos, const TTableHints& hints);
     TNodePtr BuildInputTables(TPosition pos, const TTableList& tables, bool inSubquery, TScopedStatePtr scoped);
-    TNodePtr BuildCreateTable(TPosition pos, const TTableRef& tr, bool existingOk, const TCreateTableParameters& params, TScopedStatePtr scoped);
+    TNodePtr BuildCreateTable(TPosition pos, const TTableRef& tr, bool existingOk, bool isReplace, const TCreateTableParameters& params, TScopedStatePtr scoped);
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/source.h]: Fixed

File: ydb/library/yql/sql/v1/sql_query.cpp [hunk 11]
```diff
@@ -163,15 +163,28 @@ bool TSqlQuery::Statement(TVector<TNodePtr>& blocks, const TRule_sql_stmt_core&
             }
 
             bool existingOk = false;
-            if (rule.HasBlock3()) { // IF NOT EXISTS
+            bool isReplace = false;
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/sql_query.cpp]: Fixed

File: ydb/library/yql/sql/v1/sql_query.cpp [hunk 12]
```diff
@@ -745,12 +758,19 @@ bool TSqlQuery::Statement(TVector<TNodePtr>& blocks, const TRule_sql_stmt_core&
             }
 
             bool existingOk = false;
-            if (node.HasBlock5()) { // IF NOT EXISTS
+            bool isReplace = false;
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/library/yql/sql/v1/sql_query.cpp]: Fixed

File: ydb/core/kqp/provider/yql_kikimr_gateway.h [hunk 13]
```diff
@@ -790,7 +790,7 @@ class IKikimrGateway : public TThrRefBase {
     virtual NThreading::TFuture<TTableMetadataResult> LoadTableMetadata(
         const TString& cluster, const TString& table, TLoadTableMetadataSettings settings) = 0;
 
-    virtual NThreading::TFuture<TGenericResult> CreateTable(TKikimrTableMetadataPtr metadata, bool createDir, bool existingOk = false) = 0;
+    virtual NThreading::TFuture<TGenericResult> CreateTable(TKikimrTableMetadataPtr metadata, bool createDir, bool existingOk = false, bool isReplace = false) = 0;
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/core/kqp/provider/yql_kikimr_gateway.h]: Fixed

File: ydb/core/kqp/provider/yql_kikimr_expr_nodes.json [hunk 14]
```diff
@@ -131,7 +131,8 @@
                 {"Index": 10, "Name": "Changefeeds", "Type": "TCoChangefeedList"},
                 {"Index": 11, "Name": "TableType", "Type": "TCoAtom"},
                 {"Index": 12, "Name": "Temporary", "Type": "TCoAtom"},
-                {"Index": 13, "Name": "ExistingOk", "Type": "TCoAtom"}
+                {"Index": 13, "Name": "ExistingOk", "Type": "TCoAtom"},
+                {"Index": 14, "Name": "IsReplace", "Type":  "TCoAtom"}
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/core/kqp/provider/yql_kikimr_expr_nodes.json]: Fixed

File: ydb/core/kqp/provider/yql_kikimr_exec.cpp [hunk 15]
```diff
@@ -934,10 +934,11 @@ class TKiSinkCallableExecutionTransformer : public TAsyncCallbackTransformer<TKi
             NThreading::TFuture<IKikimrGateway::TGenericResult> future;
             bool isColumn = (table.Metadata->StoreType == EStoreType::Column);
             bool existingOk = (maybeCreate.ExistingOk().Cast().Value() == "1");
+            bool isReplace = (maybeCreate.IsReplace().Cast().Value() == "1");
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/core/kqp/provider/yql_kikimr_exec.cpp]: Fixed

File: ydb/core/kqp/provider/yql_kikimr_datasink.cpp [hunk 16]
```diff
@@ -779,6 +779,7 @@ class TKikimrDataSink : public TDataProviderBase
                         ? settings.Temporary.Cast()
                         : Build<TCoAtom>(ctx, node->Pos()).Value("false").Done();
 
+                    auto isReplace = (settings.Mode.Cast().Value() == "create_or_replace");
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/core/kqp/provider/yql_kikimr_datasink.cpp]: Fixed

File: ydb/core/kqp/host/kqp_gateway_proxy.cpp [hunk 17]
```diff
@@ -444,7 +444,8 @@ class TKqpGatewayProxy : public IKikimrGateway {
         return Gateway->LoadTableMetadata(cluster, table, settings);
     }
 
-    TFuture<TGenericResult> CreateTable(TKikimrTableMetadataPtr metadata, bool createDir, bool existingOk) override {
+    TFuture<TGenericResult> CreateTable(TKikimrTableMetadataPtr metadata, bool createDir, bool existingOk, bool isReplace) override {
```
[the remark this answers is not in the document]
Comment (Alnen) [reply ydb/core/kqp/host/kqp_gateway_proxy.cpp]: Fixed

File: ydb/library/yql/sql/v1/sql_query.cpp [hunk 18]
```diff
@@ -163,57 +173,63 @@ bool TSqlQuery::Statement(TVector<TNodePtr>& blocks, const TRule_sql_stmt_core&
...
-                    rule.GetBlock3().GetToken2().GetId() == SQLv1LexerTokens::TOKEN_NOT &&
-                    rule.GetBlock3().GetToken3().GetId() == SQLv1LexerTokens::TOKEN_EXISTS
+                    rule.GetBlock4().GetToken1().GetId() == SQLv1LexerTokens::TOKEN_IF &&
+                    rule.GetBlock4().GetToken2().GetId() == SQLv1LexerTokens::TOKEN_NOT &&
+                    rule.GetBlock4().GetToken3().GetId() == SQLv1LexerTokens::TOKEN_EXISTS
                 );
             }
 
+            if (replaceIfExists && tableType != ETableType::ExternalTable) {
+                Context().Error(GetPos(rule.GetBlock2().GetToken1()))
```
Comment (nepal) [inline ydb/library/yql/sql/v1/sql_query.cpp:186]: Please add test for this in sql_ut.cpp

**Review by nepal — CHANGES_REQUESTED** [review]

**Review by UgnineSirdis — APPROVED** [review]

**Review by nepal — APPROVED** [review]
sql part lgtm

**Pull request merged**